Compilation errors

I created a .net 5 library in VS preview. I added the DNNE package (latest version. Next I added a method `public static int Add(int a, int b)`, and marked it with the UnmanagedCallersOnlyAttribute.
When I compile in VS, I get an error saying the `DNNE.BuildTasks.dll` file couldn't be loaded.

When I run `dotnet build` from the command line, I get an error from the c++ compiler, saying it can't find `libnethost.lib`.

What am I doing wrong?

Is there a (working) sample project somewhere?

I have .NET 5 preview 6 installed on windows.

Compiler output compiling `ExportingAssembly.csproj` using `dotnet build`:

```
Microsoft (R) Build Engine version 16.7.0-preview-20310-07+ee1c9fd0c for .NET
Copyright (C) Microsoft Corporation. All rights reserved.

  Determining projects to restore...
  All projects are up-to-date for restore.
  You are using a preview version of .NET. See: https://aka.ms/dotnet-core-preview
  Building native export: "C:\Program Files (x86)\Microsoft Visual Studio\2019\Professional\VC\Tools\MSVC\14.26.28801\bin\Hostx64\x86\cl.exe" /Od /LDd /TC /MT /GS /Zi /D DNNE_ASSEMBLY_NAME=ExportingAssembly /I "C:\Program Files (x86)\Microsoft Visual Studio\2019\Professional\VC\Tools\MSVC\14.26.28801\include" /I "C:\Users\matth\.nuget\packages\dnne\1.0.5\tools\platform" /I "C:\Program Files (x86)\dotnet\packs\Microsoft.NETCore.App.Host.win-x86\5.0.0-preview.6.20305.6\runtimes\win-x86\native" /I "C:\Program Files (x86)\Windows Kits\10\Include\10.0.18362.0\shared" /I "C:\Program Files (x86)\Windows Kits\10\Include\10.0.18362.0\um" /I "C:\Program Files (x86)\Windows Kits\10\Include\10.0.18362.0\ucrt" "e:\Data\Sources\Rommel\DNNE\test\ExportingAssembly\obj\Debug\net5.0\dnne\ExportingAssembly.g.c" "C:\Users\matth\.nuget\packages\dnne\1.0.5\tools\platform\platform.c"  /link /DLL /LIBPATH:"C:\Program Files (x86)\Microsoft Visual Studio\2019\Professional\VC\Tools\MSVC\14.26.28801\lib\x86" /LIBPATH:"C:\Program Files (x86)\Windows Kits\10\Lib\10.0.18362.0\um\x86" /LIBPATH:"C:\Program Files (x86)\Windows Kits\10\Lib\10.0.18362.0\ucrt\x86" "C:\Program Files (x86)\dotnet\packs\Microsoft.NETCore.App.Host.win-x86\5.0.0-preview.6.20305.6\runtimes\win-x86\native\libnethost.lib" Advapi32.lib /IGNORE:4099 /out:"e:\Data\Sources\Rommel\DNNE\test\ExportingAssembly\obj\Debug\net5.0\dnne\bin\ExportingAssemblyNE.dll" 
  Microsoft (R) C/C++ Optimizing Compiler Version 19.26.28806 for x86
  Copyright (C) Microsoft Corporation.  All rights reserved.
  
  ExportingAssembly.g.c
  platform.c
  Generating Code...
  Microsoft (R) Incremental Linker Version 14.26.28806.0
  Copyright (C) Microsoft Corporation.  All rights reserved.
  
  /out:ExportingAssembly.g.dll 
  /dll 
  /implib:ExportingAssembly.g.lib 
  /debug 
  /DLL 
  "/LIBPATH:C:\Program Files (x86)\Microsoft Visual Studio\2019\Professional\VC\Tools\MSVC\14.26.28801\lib\x86" 
  "/LIBPATH:C:\Program Files (x86)\Windows Kits\10\Lib\10.0.18362.0\um\x86" 
  "/LIBPATH:C:\Program Files (x86)\Windows Kits\10\Lib\10.0.18362.0\ucrt\x86" 
  "C:\Program Files (x86)\dotnet\packs\Microsoft.NETCore.App.Host.win-x86\5.0.0-preview.6.20305.6\runtimes\win-x86\native\libnethost.lib" 
  Advapi32.lib 
  /IGNORE:4099 
  /out:e:\Data\Sources\Rommel\DNNE\test\ExportingAssembly\obj\Debug\net5.0\dnne\bin\ExportingAssemblyNE.dll 
  ExportingAssembly.g.obj 
  platform.obj 
LINK : fatal error LNK1181: cannot open input file 'C:\Program Files (x86)\dotnet\packs\Microsoft.NETCore.App.Host.win-x86\5.0.0-preview.6.20305.6\runtimes\win-x86\native\libnethost.lib' [e:\Data\Sources\Rommel\DNNE\test\ExportingAssembly\ExportingAssembly.csproj]
C:\Users\matth\.nuget\packages\dnne\1.0.5\build\DNNE.targets(102,5): error MSB3073: The command ""C:\Program Files (x86)\Microsoft Visual Studio\2019\Professional\VC\Tools\MSVC\14.26.28801\bin\Hostx64\x86\cl.exe" /Od /LDd /TC /MT /GS /Zi /D DNNE_ASSEMBLY_NAME=ExportingAssembly /I "C:\Program Files (x86)\Microsoft Visual Studio\2019\Professional\VC\Tools\MSVC\14.26.28801\include" /I "C:\Users\matth\.nuget\packages\dnne\1.0.5\tools\platform" /I "C:\Program Files (x86)\dotnet\packs\Microsoft.NETCore.App.Host.win-x86\5.0.0-preview.6.20305.6\runtimes\win-x86\native" /I "C:\Program Files (x86)\Windows Kits\10\Include\10.0.18362.0\shared" /I "C:\Program Files (x86)\Windows Kits\10\Include\10.0.18362.0\um" /I "C:\Program Files (x86)\Windows Kits\10\Include\10.0.18362.0\ucrt" "e:\Data\Sources\Rommel\DNNE\test\ExportingAssembly\obj\Debug\net5.0\dnne\ExportingAssembly.g.c" "C:\Users\matth\.nuget\packages\dnne\1.0.5\tools\platform\platform.c"  /link /DLL /LIBPATH:"C:\Program Files (x86)\Microsoft Visual Studio\2019\Professional\VC\Tools\MSVC\14.26.28801\lib\x86" /LIBPATH:"C:\Program Files (x86)\Windows Kits\10\Lib\10.0.18362.0\um\x86" /LIBPATH:"C:\Program Files (x86)\Windows Kits\10\Lib\10.0.18362.0\ucrt\x86" "C:\Program Files (x86)\dotnet\packs\Microsoft.NETCore.App.Host.win-x86\5.0.0-preview.6.20305.6\runtimes\win-x86\native\libnethost.lib" Advapi32.lib /IGNORE:4099 /out:"e:\Data\Sources\Rommel\DNNE\test\ExportingAssembly\obj\Debug\net5.0\dnne\bin\ExportingAssemblyNE.dll" " exited with code 2. [e:\Data\Sources\Rommel\DNNE\test\ExportingAssembly\ExportingAssembly.csproj]

Build FAILED.

LINK : fatal error LNK1181: cannot open input file 'C:\Program Files (x86)\dotnet\packs\Microsoft.NETCore.App.Host.win-x86\5.0.0-preview.6.20305.6\runtimes\win-x86\native\libnethost.lib' [e:\Data\Sources\Rommel\DNNE\test\ExportingAssembly\ExportingAssembly.csproj]
C:\Users\matth\.nuget\packages\dnne\1.0.5\build\DNNE.targets(102,5): error MSB3073: The command ""C:\Program Files (x86)\Microsoft Visual Studio\2019\Professional\VC\Tools\MSVC\14.26.28801\bin\Hostx64\x86\cl.exe" /Od /LDd /TC /MT /GS /Zi /D DNNE_ASSEMBLY_NAME=ExportingAssembly /I "C:\Program Files (x86)\Microsoft Visual Studio\2019\Professional\VC\Tools\MSVC\14.26.28801\include" /I "C:\Users\matth\.nuget\packages\dnne\1.0.5\tools\platform" /I "C:\Program Files (x86)\dotnet\packs\Microsoft.NETCore.App.Host.win-x86\5.0.0-preview.6.20305.6\runtimes\win-x86\native" /I "C:\Program Files (x86)\Windows Kits\10\Include\10.0.18362.0\shared" /I "C:\Program Files (x86)\Windows Kits\10\Include\10.0.18362.0\um" /I "C:\Program Files (x86)\Windows Kits\10\Include\10.0.18362.0\ucrt" "e:\Data\Sources\Rommel\DNNE\test\ExportingAssembly\obj\Debug\net5.0\dnne\ExportingAssembly.g.c" "C:\Users\matth\.nuget\packages\dnne\1.0.5\tools\platform\platform.c"  /link /DLL /LIBPATH:"C:\Program Files (x86)\Microsoft Visual Studio\2019\Professional\VC\Tools\MSVC\14.26.28801\lib\x86" /LIBPATH:"C:\Program Files (x86)\Windows Kits\10\Lib\10.0.18362.0\um\x86" /LIBPATH:"C:\Program Files (x86)\Windows Kits\10\Lib\10.0.18362.0\ucrt\x86" "C:\Program Files (x86)\dotnet\packs\Microsoft.NETCore.App.Host.win-x86\5.0.0-preview.6.20305.6\runtimes\win-x86\native\libnethost.lib" Advapi32.lib /IGNORE:4099 /out:"e:\Data\Sources\Rommel\DNNE\test\ExportingAssembly\obj\Debug\net5.0\dnne\bin\ExportingAssemblyNE.dll" " exited with code 2. [e:\Data\Sources\Rommel\DNNE\test\ExportingAssembly\ExportingAssembly.csproj]
    0 Warning(s)
    2 Error(s)

Time Elapsed 00:00:01.22
```

--- a/readme.md
+++ b/readme.md
@@ -89,6 +89,8 @@ The `preload_runtime()` function can be used to preload the runtime. This may be
     ```
 
 1) Build the managed project to generate the native binary. The native binary will have a `NE` suffix and the system extension for dynamic/shared native libraries (i.e. `.dll`, `.so`, `.dylib`).
+    * The [Runtime Identifier (RID)](https://docs.microsoft.com/dotnet/core/rid-catalog) is used to target a specific SDK.
+    * For example, on Windows the `--runtime` flag can be used to target `win-x86` or `win-x64`.
 
 1) Deploy the native binary, managed assembly and associated `*.json` files for consumption from a native process.
 

--- a/src/dnne-gen/dnne-gen.csproj
+++ b/src/dnne-gen/dnne-gen.csproj
@@ -15,7 +15,7 @@
 
   <PropertyGroup>
     <PackageId>DNNE</PackageId>
-    <Version>1.0.5</Version>
+    <Version>1.0.6</Version>
     <Authors>AaronRobinsonMSFT</Authors>
     <Owners>AaronRobinsonMSFT</Owners>
     <Description>Package used to generated native exports for .NET assemblies.</Description>

--- a/src/msbuild/DNNE.BuildTasks/Windows.cs
+++ b/src/msbuild/DNNE.BuildTasks/Windows.cs
@@ -92,11 +92,12 @@ namespace DNNE.BuildTasks
 
         private static bool Is64BitTarget(string arch)
         {
-            return arch switch
+            return arch.ToLower() switch
             {
-                "AnyCPU" => IntPtr.Size == 8,
                 "x64" => true,
-                _ => false,
+                "amd64" => true,
+                "x86" => false,
+                _ => IntPtr.Size == 8,
             };
         }
 

--- a/src/msbuild/DNNE.BuildTasks/macOS.cs
+++ b/src/msbuild/DNNE.BuildTasks/macOS.cs
@@ -57,9 +57,13 @@ namespace DNNE.BuildTasks
 
         private static bool Is64BitTarget(string arch)
         {
-            return arch switch
+            return arch.ToLower() switch
             {
                 "AnyCPU" => true,
+                "msil" => true,
+                "Neutral" => true,
+                "x64" => true,
+                "amd64" => true,
                 _ => throw new Exception("Unknown architecture. Only support 'AnyCPU'."),
             };
         }

--- a/src/msbuild/DNNE.targets
+++ b/src/msbuild/DNNE.targets
@@ -80,7 +80,9 @@ DNNE.targets
 
     <PropertyGroup>
       <DnneAssemblyName>$(TargetName)</DnneAssemblyName>
-      <DnneNetHostDir>$(NetCoreTargetingPackRoot)/Microsoft.NETCore.App.Host.$(NETCoreSdkRuntimeIdentifier)/$(BundledNETCoreAppPackageVersion)/runtimes/$(NETCoreSdkRuntimeIdentifier)/native</DnneNetHostDir>
+      <DnneRuntimeIdentifier>$(RuntimeIdentifier)</DnneRuntimeIdentifier>
+      <DnneRuntimeIdentifier Condition="'$(DnneRuntimeIdentifier)'==''">$(NETCoreSdkRuntimeIdentifier)</DnneRuntimeIdentifier>
+      <DnneNetHostDir>$(NetCoreTargetingPackRoot)/Microsoft.NETCore.App.Host.$(DnneRuntimeIdentifier)/$(BundledNETCoreAppPackageVersion)/runtimes/$(DnneRuntimeIdentifier)/native</DnneNetHostDir>
       <DnneCompiledToBinPath>$(DnneGeneratedBinPath)/$(DnneNativeExportsBinaryName)$(DnneNativeBinaryExt)</DnneCompiledToBinPath>
       <__DnneGeneratedSourceFile>@(DnneGeneratedSourceFile)</__DnneGeneratedSourceFile>
     </PropertyGroup>
@@ -92,7 +94,7 @@ DNNE.targets
         Source="$([MSBuild]::NormalizePath($(__DnneGeneratedSourceFile)))"
         OutputName="$(DnneNativeExportsBinaryName)$(DnneNativeBinaryExt)"
         OutputPath="$([MSBuild]::NormalizePath($(DnneGeneratedBinPath)))"
-        Architecture="$(Platform)"
+        Architecture="$(TargetedSDKArchitecture)"
         Configuration="$(Configuration)">
       <Output TaskParameter="Command" PropertyName="CompilerCmd" />
       <Output TaskParameter="CommandArguments" PropertyName="CompilerArgs" />

--- a/test/ExportingAssembly/ExportingAssembly.csproj
+++ b/test/ExportingAssembly/ExportingAssembly.csproj
@@ -10,4 +10,8 @@
     <PackageReference Include="DNNE" Version="1.*" />
   </ItemGroup>
 
+  <PropertyGroup Condition="'$(LocalNuPkg)'!=''">
+    <RestoreSources>$(LocalNuPkg)</RestoreSources>
+  </PropertyGroup>
+
 </Project>

--- a/test/ExportingAssembly/nuget.config
+++ b/test/ExportingAssembly/nuget.config
@@ -1,6 +1,0 @@
-<?xml version="1.0" encoding="utf-8"?>
-<configuration>
-    <packageSources>
-        <add key="LOCAL_TESTING" value="" />
-    </packageSources>
-</configuration>

